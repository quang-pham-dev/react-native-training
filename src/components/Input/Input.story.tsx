import React from 'react'
import {storiesOf} from '@storybook/react-native'
import {action} from '@storybook/addon-actions'
import {StoryScreen, Story, UseCase} from '../../../storybook/views'
import {text} from '@storybook/addon-knobs'

import {Icons} from '@themes'
import Input, {InputType} from './index'

declare let module: NodeModule

storiesOf('Input', module)
  .addDecorator(fn => <StoryScreen>{fn()}</StoryScreen>)
  .add('Type normal', () => (
    <Story>
      <UseCase text={'normal'}>
        <Input
          label={text('label', 'Username')}
          placeholder={text('placeholder', 'Enter your username')}
          icon={Icons.filledProfile}
          value={''}
          type={InputType.NORMAL}
          onChange={action('onChange text')}
        />
      </UseCase>
    </Story>
  ))
  .add('Type password', () => (
    <Story>
      <UseCase text={'password'}>
        <Input
          label={text('label', 'Password')}
          placeholder={text('placeholder', 'Enter your password')}
          value={''}
          type={InputType.PASSWORD}
          onChange={action('onChange text')}
        />
      </UseCase>
    </Story>
  ))
