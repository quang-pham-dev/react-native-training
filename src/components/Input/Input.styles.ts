import styled from 'styled-components/native'

// Styles
import FlexStyled from '@components/Flex/Flex.styles'

// Themes
import {Fonts, Colors} from '@themes'

interface IconWrapperProps {
  mRight?: number
  mBottom?: number
}
interface TextInputProps {
  pLeft?: number
  pRight?: number
}

interface IInputStyled {
  Wrapper: typeof Wrapper
  LabelStyled: typeof LabelStyled
  InputField: typeof InputField
  IconEyeWrapper: typeof IconEyeWrapper
}

const Wrapper = styled(FlexStyled.RowCenterHorizontal)`
  border-bottom-width: 1px;
  border-bottom-color: ${Colors.palette.borderGray};
  width: 100%;

  ${props => props.pHorizontal && `padding-horizontal: ${props.pHorizontal}px`};
  ${props => props.pRight && `padding-right: ${props.pRight}px`};
`

const LabelStyled = styled.Text`
 font-size: ${Fonts.size.small}px;
 line-height:  ${Fonts.lineHeight.xs}px
 font-family: ${Fonts.type.interRegular};
 color:${Colors.palette.gray};
`

const InputField = styled.TextInput<TextInputProps>`
  flex: 1;
  padding-left: ${props => (props.pLeft ? props.pLeft : 0)}px;
  padding-right: ${props => (props.pRight ? props.pRight : 5)}px;
  padding-vertical: 15px;
  font-family: ${Fonts.type.interMedium};
  font-size: ${Fonts.size.normal}px;
  line-height: ${Fonts.lineHeight.md}px;
  color: ${Colors.palette.black};
`

const IconEyeWrapper = styled.TouchableOpacity<IconWrapperProps>`
  ${props => props.mRight && `margin-right: ${props.mRight}px;`}
  ${props => props.mBottom && `margin-bottom: ${props.mBottom}px;`}
`

const InputStyled: IInputStyled = {
  Wrapper,
  LabelStyled,
  InputField,
  IconEyeWrapper,
}

export default InputStyled
