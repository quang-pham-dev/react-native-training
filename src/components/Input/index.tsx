import React, {memo, useCallback, useState} from 'react'
import {
  ImageSourcePropType,
  NativeSyntheticEvent,
  TextInputChangeEventData,
} from 'react-native'

// LIBS
import isEqual from 'react-fast-compare'

// Styles
import IconStyled from '@components/Icon/Icon.styles'
import FlexStyled from '@components/Flex/Flex.styles'
import InputStyled from './Input.styles'

// Themes
import {Icons} from '@themes'

export enum InputType {
  NORMAL = 'normal',
  EMAIL = 'email',
  PHONE = 'phone',
  NUMBER = 'number',
  PASSWORD = 'password',
}

export enum KeyboardType {
  DEFAULT = 'default',
  NUMERIC = 'numeric',
  EMAIL_ADDRESS = 'email-address',
  ASCII_CAPABLE = 'ascii-capable',
  NUMBER_PUNCTUATION = 'numbers-and-punctuation',
  URL = 'url',
  NUMBER_PAD = 'number-pad',
  PHONE_PAD = 'phone-pad',
  NAME_PHONE_PAD = 'name-phone-pad',
  DECIMAL_PAD = 'decimal-pad',
  TWITTER = 'twitter',
  WEB_SEARCH = 'web-search',
  VISIBLE_PASSWORD = 'visible-password',
}

type InputProps = {
  type?: InputType
  value: string
  label?: string
  placeholder?: string
  icon?: ImageSourcePropType
  keyboardType?: KeyboardType
  autoFocus?: boolean
  onBlur?: () => void
  onChange: (value: string) => void | undefined
  onEndEditing?: (
    event: NativeSyntheticEvent<TextInputChangeEventData>,
    type: InputType,
  ) => void
  onClearInput?: () => void
  onSubmitEditing?: () => void
  inputProps?: {
    [key: string]: string | number
  }
}

const Input = ({
  type = InputType.NORMAL,
  value,
  label,
  placeholder,
  icon,
  keyboardType = KeyboardType.DEFAULT,
  onChange,
  onBlur,
  onSubmitEditing,
  inputProps,
}: InputProps) => {
  const isPassword = type === InputType.PASSWORD
  const [displayValue, setDisplayValue] = useState(value)
  const [showPassword, setShowPassword] = useState<boolean>(false)

  const handleTextChange = useCallback(
    (text: string) => {
      setDisplayValue(text)
      onChange(text)
    },
    [onChange, setDisplayValue],
  )

  const handleTogglePassword = (): void => setShowPassword(!showPassword)

  return (
    <FlexStyled.Default flex={1}>
      {label && <InputStyled.LabelStyled>{label}</InputStyled.LabelStyled>}

      <InputStyled.Wrapper>
        <InputStyled.InputField
          value={displayValue}
          onChangeText={handleTextChange}
          onBlur={onBlur}
          autoCapitalize="none"
          underlineColorAndroid="transparent"
          onSubmitEditing={onSubmitEditing}
          placeholder={placeholder}
          keyboardType={keyboardType}
          secureTextEntry={!showPassword}
          {...inputProps}
        />
        {isPassword && (
          <InputStyled.IconEyeWrapper onPress={handleTogglePassword}>
            <IconStyled source={showPassword ? Icons.eyeFill : Icons.eyeOff} />
          </InputStyled.IconEyeWrapper>
        )}
        {icon && <IconStyled source={icon} />}
      </InputStyled.Wrapper>
    </FlexStyled.Default>
  )
}

export default memo(Input, isEqual)
