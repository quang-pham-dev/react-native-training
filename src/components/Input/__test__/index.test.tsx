import React from 'react'

// LIBS
import {render} from '@testing-library/react-native'

// Components
import Input from '@components/Input'

describe('TextInput Component', () => {
  const props = {
    value: 'user name',
    labelStyle: {},
    placeholder: 'Enter your user name',
    textInputStyles: {},
    secureTextEntry: false,
    autoFocus: false,
    onChange: jest.fn(),
    onSubmitEditing: jest.fn(),
    onBlur: jest.fn(),
  }

  const tree = render(<Input {...props} />)
  test('should render correctly', () => {
    const component = tree.toJSON()
    expect(component).toMatchSnapshot()
  })

  test('should render with label correctly', () => {
    const newProps = {
      ...props,
      label: 'Username',
    }
    const component = render(<Input {...newProps} />).toJSON()
    expect(component).toMatchSnapshot()
  })
  test('should render placeholder', () => {
    const newProps = {
      ...props,
      placeholder: 'Enter your username',
    }
    const component = render(<Input {...newProps} />).toJSON()
    expect(component).toMatchSnapshot()
  })
})
