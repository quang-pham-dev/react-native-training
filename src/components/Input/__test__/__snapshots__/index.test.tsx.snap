// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Component should render correctly 1`] = `
<View
  flex={1}
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "#E7E8EA",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "width": "100%",
        },
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      keyboardType="default"
      onBlur={[MockFunction]}
      onChangeText={[Function]}
      onSubmitEditing={[MockFunction]}
      placeholder="Enter your user name"
      secureTextEntry={true}
      style={
        Array [
          Object {
            "color": "#1D1E20",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontFamily": "Inter-Medium",
            "fontSize": 15,
            "lineHeight": 16,
            "paddingLeft": 0,
            "paddingRight": 5,
            "paddingVertical": 15,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value="user name"
    />
  </View>
</View>
`;

exports[`TextInput Component should render placeholder 1`] = `
<View
  flex={1}
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "#E7E8EA",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "width": "100%",
        },
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      keyboardType="default"
      onBlur={[MockFunction]}
      onChangeText={[Function]}
      onSubmitEditing={[MockFunction]}
      placeholder="Enter your username"
      secureTextEntry={true}
      style={
        Array [
          Object {
            "color": "#1D1E20",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontFamily": "Inter-Medium",
            "fontSize": 15,
            "lineHeight": 16,
            "paddingLeft": 0,
            "paddingRight": 5,
            "paddingVertical": 15,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value="user name"
    />
  </View>
</View>
`;

exports[`TextInput Component should render with label correctly 1`] = `
<View
  flex={1}
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#8F959E",
          "fontFamily": "Inter-Regular",
          "fontSize": 13,
          "lineHeight": 14,
        },
      ]
    }
  >
    Username
  </Text>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "#E7E8EA",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "width": "100%",
        },
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      keyboardType="default"
      onBlur={[MockFunction]}
      onChangeText={[Function]}
      onSubmitEditing={[MockFunction]}
      placeholder="Enter your user name"
      secureTextEntry={true}
      style={
        Array [
          Object {
            "color": "#1D1E20",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontFamily": "Inter-Medium",
            "fontSize": 15,
            "lineHeight": 16,
            "paddingLeft": 0,
            "paddingRight": 5,
            "paddingVertical": 15,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value="user name"
    />
  </View>
</View>
`;
